### ISSUE TITLE
K.dot behaves differently on Theano and Tensorflow

### ISSUE THREAD

**Issue by the-moliver** [fchollet/keras#2153]
This issue is that Theano has broadcasting for matrix multiplication and Tensorflow does not.
So while ```T.dot``` and ```tf.matmul``` are the same when the inputs are 2D, Theano happily takes 3D inputs while Tensorflow does not.
This is the issue breaking maxout in tensorflow: https://github.com/fchollet/keras/issues/2149
This seems best resolved by forcing ```K.dot``` to only take 2D input and then using ```K.batch_dot``` for things like maxout. What do you all think?

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by the-moliver** [comment]
Or wait for the tensorflow folks to fix https://github.com/tensorflow/tensorflow/issues/216

**Comment by fchollet** [comment]
The strange about this is how we apparently didn't have a unit test for
maxout. How did that happen?

I am fine with only using `dot` for 2D matrices, and having `batch_dot`
otherwise.

**Comment by shampool** [comment]
Instead of introducing two functions and change the API, why not just make 'dot' for 2D as a special case for K.dot just like theano did.

**Comment by fchollet** [comment]
So what's your proposed version of Maxout based on `tf.batch_dot`?

**Comment by the-moliver** [comment]
```tf.batch_dot``` doesn't seem like the best choice for maxout actually.  I got a version working using 2d matrix multiply and reshapes however.  I'll push it soon, likely Sunday. 

I don't understand what you mean shampool. The issue is that tensorflow does not support broadcasting (yet).  So the proposal was to limit the keras backend to the tensorflow way of doing things (and then remove the limitataion when TF changes)

**Comment by shampool** [comment]
I mean `K.dot ` backend for tensorflow can be made to mimic the  behavior of `T.dot`. 
For example, add a condition (dimension of input == 2) to determine it should execute `tf.matmul `or `tf.batch_dot`.

**Comment by fchollet** [comment]
FYI I have fixed this in the TF backend earlier this afternoon (see keras-1
branch).

**Comment by the-moliver** [comment]
That fix will work in one case,  which admittedly is probably the most common one,  but the behavior of theano's dot function is far more general. This could lead to future incompatibility between the two backends.  Maybe fine for now,  until tensorflow supports proper broadcasting?

**Comment by the-moliver** [comment]
Also I believe that fix can be done without a loop by reshaping before and after a matrix multiply.  I think that would be faster?

**Comment by SarthakMehta** [comment]
replace following with existing method for last dimension dot product

def dot(self,x,y):
        if K.ndim(x) is not None and (K.ndim(x) > 2 or K.ndim(y) > 2):
            x_shape = []
            for i, s in zip(K.int_shape(x), tf.unstack(tf.shape(x))):
                if i is not None:
                    x_shape.append(i)
                else:
                    x_shape.append(s)
            x_shape = tuple(x_shape)
            y_shape = []
            for i, s in zip(K.int_shape(y), tf.unstack(tf.shape(y))):
                if i is not None:
                    y_shape.append(i)
                else:
                    y_shape.append(s)
            y_shape = tuple(y_shape)
            xt = tf.reshape(x, [-1, x_shape[-1]])
            yt = tf.reshape(tf.transpose(y), [y_shape[-1], -1])
            
            return tf.reshape(tf.matmul(xt, yt),x_shape[:-1] + y_shape[-1:])
        if K.is_sparse(x):
            out = tf.sparse_tensor_dense_matmul(x, y)
        else:
            out = tf.matmul(x, y)
        return out

**Comment by AlexGidiotis** [comment]
In order to get the functionality:
`K.dot(x, kernel)`
with Tensorflow backend you can try:
`K.squeeze(K.dot(x, K.expand_dims(kernel)), axis=-1)`
